File: backend/services/documentsService.js
const doc = require('pdfkit');
const db = require('../models');

class DocumentsService {
  parseBase64Blob(blobInput) {
    if (!blobInput) {
      throw new Error('Document blob is required (base64)');
    }

    if (Buffer.isBuffer(blobInput)) {
      return blobInput;
    }

    if (typeof blobInput !== 'string') {
      throw new Error('Document blob must be a base64 string');
    }

    const cleaned = blobInput.startsWith('data:')
      ? blobInput.substring(blobInput.indexOf(',') + 1)
      : blobInput;

    if (!cleaned || cleaned.trim().length === 0) {
      throw new Error('Document blob is empty');
    }

    const normalized = cleaned.replace(/\s/g, '');
    if (!/^[A-Za-z0-9+/]*={0,2}$/.test(normalized)) {
      throw new Error('Document blob must be valid base64');
    }

    return Buffer.from(normalized, 'base64');
  }

  async getDocumentsByRadniNalog(radniNalogId) {
    try {
      const documents = await db.document.findAll({
        where: {
          radni_nalog_id: radniNalogId
        }
      });

      return documents;
    } catch (error) {
      throw new Error(`Error fetching documents: ${error.message}`);
    }
  }

  async addDocument(radniNalogId, data) {
    try {
      // Verify radni nalog exists
      const nalog = await db.radni_nalog.findByPk(radniNalogId);
      if (!nalog) {
        throw new Error('Radni nalog not found');
      }

      if (!data.name) {
        throw new Error('Document name is required');
      }

      if (!data.blob && data.url === undefined) {
        throw new Error('Document blob is required (base64)');
      }

      if (data.url !== undefined && typeof data.url !== 'string') {
        throw new Error('Document url must be a string');
      }
      
      let documentDate = {
          name: data.name,
          radni_nalog_id: radniNalogId
        };

      if(data.blob) {
        let blobValue = this.parseBase64Blob(data.blob);
        documentDate.blob = blobValue;
      }
      else {
        documentDate.url = data.url;
      }

      console.log('Adding document with data:', documentDate);

      return await db.document.create(documentDate);
    } catch (error) {
      throw new Error(`Error adding document: ${error.message}`);
    }
  }

  async getDocumentById(radniNalogId, documentId) {
    try {
      const document = await db.document.findOne({
        where: {
          id: documentId,
          radni_nalog_id: radniNalogId
        }
      });

      if (!document) {
        throw new Error('Document not found');
      }

      return document;
    } catch (error) {
      throw new Error(`Error fetching document: ${error.message}`);
    }
  }

  async deleteDocument(documentId) {
    try {
      const document = await db.document.findByPk(documentId);

      if (!document) {
        throw new Error('Document not found');
      }

      await document.destroy();
      return document;
    } catch (error) {
      throw new Error(`Error deleting document: ${error.message}`);
    }
  }
}

module.exports = new DocumentsService();

File: JavaSimpleBackend/src/main/java/com/atesti/service/DocumentsService.java
package com.atesti.service;

import com.atesti.dto.DocumentRequest;
import com.atesti.entity.Document;
import com.atesti.entity.RadniNalog;
import com.atesti.exception.BadRequestException;
import com.atesti.exception.ResourceNotFoundException;
import com.atesti.repository.DocumentRepository;
import com.atesti.repository.RadniNalogRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Base64;
import java.util.List;

@Service
@RequiredArgsConstructor
public class DocumentsService {

    private final DocumentRepository documentRepository;
    private final RadniNalogRepository radniNalogRepository;

    public List<Document> getDocumentsByRadniNalog(Long radniNalogId) {
        return documentRepository.findByRadniNalogId(radniNalogId);
    }

    public Document addDocument(Long radniNalogId, DocumentRequest request) {
        RadniNalog nalog = radniNalogRepository.findById(radniNalogId)
                .orElseThrow(() -> new BadRequestException("Error adding document: Radni nalog not found"));

        if (request.getName() == null || request.getName().isBlank()) {
            throw new BadRequestException("Error adding document: Document name is required");
        }

        if ((request.getBlob() == null || request.getBlob().isBlank()) && request.getUrl() == null) {
            throw new BadRequestException("Error adding document: Document blob is required (base64)");
        }

        Document.DocumentBuilder builder = Document.builder()
                .name(request.getName())
                .radniNalog(nalog);

        if (request.getBlob() != null && !request.getBlob().isBlank()) {
            byte[] blobBytes = parseBase64Blob(request.getBlob());
            builder.blob(blobBytes);
        } else {
            builder.url(request.getUrl());
        }

        return documentRepository.save(builder.build());
    }

    public Document getDocumentById(Long radniNalogId, Long documentId) {
        return documentRepository.findByIdAndRadniNalogId(documentId, radniNalogId)
                .orElseThrow(() -> new ResourceNotFoundException("Document not found"));
    }

    public Document deleteDocument(Long documentId) {
        Document document = documentRepository.findById(documentId)
                .orElseThrow(() -> new ResourceNotFoundException("Document not found"));

        documentRepository.delete(document);
        return document;
    }

    private byte[] parseBase64Blob(String blobInput) {
        if (blobInput == null || blobInput.isBlank()) {
            throw new BadRequestException("Error adding document: Document blob is required (base64)");
        }

        String cleaned = blobInput;
        if (cleaned.startsWith("data:")) {
            int commaIndex = cleaned.indexOf(',');
            if (commaIndex >= 0) {
                cleaned = cleaned.substring(commaIndex + 1);
            }
        }

        cleaned = cleaned.replaceAll("\\s", "");

        if (cleaned.isEmpty()) {
            throw new BadRequestException("Error adding document: Document blob is empty");
        }

        if (!cleaned.matches("^[A-Za-z0-9+/]*={0,2}$")) {
            throw new BadRequestException("Error adding document: Document blob must be valid base64");
        }

        return Base64.getDecoder().decode(cleaned);
    }
}
